lib/IRGen/GenCall.cpp

```diff
@@ -4282,32 +4295,56 @@ Callee irgen::getCFunctionPointerCallee(IRGenFunction &IGF,
   auto authInfo =
     PointerAuthInfo::forFunctionPointer(IGF.IGM, calleeInfo.OrigFnType);
 
-  FunctionPointer fn(fnPtr, authInfo, sig);
+  FunctionPointer fn(FunctionPointer::KindTy::Function, fnPtr, authInfo, sig);
 
   return Callee(std::move(calleeInfo), fn);
 }
 
-FunctionPointer
-FunctionPointer::forDirect(IRGenModule &IGM, llvm::Constant *fnPtr,
-                           CanSILFunctionType fnType) {
-  return forDirect(fnPtr, IGM.getSignature(fnType));
+FunctionPointer FunctionPointer::forDirect(IRGenModule &IGM,
+                                           FunctionPointer::KindTy kind,
+                                           llvm::Constant *fnPtr,
+                                           CanSILFunctionType fnType) {
+  if (kind == KindTy::AsyncFunctionPointer) {
+    assert(fnType->getRepresentation() != SILFunctionTypeRepresentation::Thick);
+  }
+  return forDirect(kind, fnPtr, IGM.getSignature(fnType));
```

Comment:
Something like this?

```
// Have to dig the size out of the AsyncFunctionPointer so that it can be rewritten downstream
%afp = `AsyncFunctionPointer(callee)`
%size_addr = getelementptr %afp, 0, 1
%size = load %size_addr
%context = call @swift_task_alloc(%task, %size)
// But we don't have to dig the function pointer out of the AsyncFunctionPointer because we already know what it is
call @callee(%task, %executor, %context)
```